src/main/java/uk/gov/ons/ctp/response/sample/scheduled/distribution/SampleUnitDistributor.java

```diff
@@ -1,170 +1,121 @@
     }
+  }
 
-    distInfo.setSampleUnitsSucceeded(successes);
-    distInfo.setSampleUnitsFailed(failures);
+  private void processJob(CollectionExerciseJob job) {
+    List<SampleUnit> mappedSampleUnits =
+        getMappedSampleUnits(job.getSampleSummaryId(), job.getCollectionExerciseId().toString());
 
-    log.info("SampleUnitsDistributor sleeping");
-    return distInfo;
-  }
+    boolean hasErrors = false;
 
-  private SampleAttributes mapSampleAttributes(
-      uk.gov.ons.ctp.response.sample.domain.model.SampleAttributes sampleAttributes) {
-    SampleAttributes mappedSampleAttributes = new SampleAttributes();
-    SampleAttributes.Builder<Void> sampleAttributesBuilder =
-        mappedSampleAttributes.newCopyBuilder();
-    for (Map.Entry<String, String> attribute : sampleAttributes.getAttributes().entrySet()) {
-      sampleAttributesBuilder
-          .addEntries()
-          .withKey(attribute.getKey())
-          .withValue(attribute.getValue());
+    for (SampleUnit msu : mappedSampleUnits) {
+      try {
+        sampleUnitSender.sendSampleUnit(msu);
+      } catch (CTPException e) {
+        hasErrors = true;
+        log.error(
+            "Failed to send a sample unit to queue and update state with ID: {}", msu.getId(), e);
+      }
     }
-    return sampleAttributesBuilder.build();
-  }
 
-  /**
-   * Sends SampleUnit to the CollectionExercise queue
-   *
-   * @param sampleUnit sample unit to be mapped
-   * @param mappedSampleUnit mapped sample unit to be sent
-   * @throws CTPException if transition issue
-   */
-  private void sendSampleUnitToCollectionExerciseQueue(
-      SampleUnit sampleUnit,
-      uk.gov.ons.ctp.response.sampleunit.definition.SampleUnit mappedSampleUnit)
-      throws CTPException {
-    transitionSampleUnitStateFromDeliveryEvent(sampleUnit);
-    sampleUnitPublisher.send(mappedSampleUnit);
+    if (!hasErrors) {
+      job.setJobComplete(true);
+      collectionExerciseJobRepository.saveAndFlush(job);
+    }
   }
 
-  private SampleUnit transitionSampleUnitStateFromDeliveryEvent(SampleUnit sampleUnit)
-      throws CTPException {
-    SampleUnitDTO.SampleUnitState newState =
-        sampleUnitStateTransitionManager.transition(
-            sampleUnit.getState(), SampleUnitDTO.SampleUnitEvent.DELIVERING);
-    sampleUnit.setState(newState);
-    sampleUnitRepository.saveAndFlush(sampleUnit);
-    return sampleUnit;
+  private List<SampleUnit> getMappedSampleUnits(UUID sampleSummaryId, String collectionExerciseId) {
+
+    List<SampleUnit> mappedSampleUnits = new LinkedList<>();
+
+    SampleSummary sampleSummary = sampleSummaryRepository.findById(sampleSummaryId);
+
+    if (sampleSummary.getState() == SampleState.ACTIVE) {
+      try (Stream<uk.gov.ons.ctp.response.sample.domain.model.SampleUnit> sampleUnits =
+          sampleUnitRepository.findBySampleSummaryFKAndState(
+              sampleSummary.getSampleSummaryPK(), SampleUnitState.PERSISTED)) {
+        sampleUnits.forEach(
```

Comment:
the idea was that we could eagerly convert the stream to a mapped list, so that we can delegate the sending to the queue (which takes much longer) to an async thread if it's taking an excessively long time